### DISCUSSION TITLE
Empirical Mode Decomposition

### DISCUSSION THREAD

**Discussion by Fakasht10** [MilesCranmer/PySR#321] [Q&A] [answered]
Reading the paper I included the link for below I was wondering why can we not reconstruct the Empirical Mode Decomposition using elementary functions with symbolic regression? Where each term in the formula represents a mode with a given frequency in order of high to low. The trend line would be picked up as well by PySR. Empirical Mode Decomposition is often used for damage detection by comparing time history response of new and used and possibly damaged system at various points on the structure. The damage per each mode of EMD would simply be the changing of parameters of adding an extra term to the expression in principle.
Regards.
Aleksey

A Symbolic Regression Based Residual Useful Life
Model for Slewing Bearings

https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjdweGtzcj-AhXtFFkFHU3ADy0QFnoECBwQAQ&url=https%3A%2F%2Fieeexplore.ieee.org%2Fiel7%2F6287639%2F6514899%2F08723621.pdf&usg=AOvVaw2CJtMmcJsL_IiuUAHoUUGY

**Comment by MilesCranmer** [comment]
Hi Aleksey,
I’m not familiar with this paper, what equation specifically do you want to reconstruct?
Best,
Miles

**Reply by Fakasht10** [reply]
say we want to use pyemd example here: can us SR to get the function itself using the simple case from example ("Input signal: $S(t)=cos(22\pi t^2) + 6t^2$")

or we can try to use SR on each of the IMS it generates, so I wanted to restrict the choise of functions to that from 1DOF normal vibration.

https://pyemd.readthedocs.io/en/latest/examples.html

**Reply by MilesCranmer** [reply]
That equation should be discoverable out of the box if you give it those operators. Did it not work? Also what is IMS?

Could you be more specific about your desired constraint?

**Reply by Fakasht10** [reply]
Apologies I meant to say IMFs (Intrinsic mode function) if you run this example from PyBMD, you will see their simple test function S(t) be decomposed in to tep IMF one that has the trend line and one that has the frequency content. PySR discovered these, but the idea is to apply it to a more complicated function or real world data and I am just trying to narrow the choice of countian types for PySR to chose from. Also I tried to find symbolic expressions for individual IMFs and that's where I ran in to nested sin, exp and cos functions. I will try to rerun it with constraints you suggested in the other thread.


from PyEMD import EMD
import numpy  as np
import pylab as plt

# Define signal
t = np.linspace(0, 1, 200)
s = np.cos(11*2*np.pi*t*t) + 6*t*t

# Execute EMD on signal
IMF = EMD().emd(s,t)
N = IMF.shape[0]+1

# Plot results
plt.subplot(N,1,1)
plt.plot(t, s, 'r')
plt.title("Input signal: $S(t)=cos(22\pi t^2) + 6t^2$")
plt.xlabel("Time [s]")

for n, imf in enumerate(IMF):
    plt.subplot(N,1,n+2)
    plt.plot(t, imf, 'g')
    plt.title("IMF "+str(n+1))
    plt.xlabel("Time [s]")

plt.tight_layout()
plt.savefig('simple_example')
plt.show()

**Reply by MilesCranmer** [reply]
I see, thanks. Yes I would try `nested_constraints` first. If that does not solve it then I would try a custom loss function with your desired functional form as the `full_objective` parameter. I give an example of this here: https://astroautomata.com/PySR/examples/#9-custom-objectives

**Reply by Fakasht10** [reply]
thank you constraints worked well! I tried slightly more complex expression and ran in to problems in the loss not going to zero.
Is there another way to speed up loss going down without setting niterations very large?

```py
X = np.random.rand(1000, 1)
y = 2 * np.cos(60 * X[:, 0]* X[:, 0]) * np.sin(60 * X[:, 0]) + 1 * X[:, 0]* X[:, 0] - 0.5

default_pysr_params = dict(
    populations=30,
    model_selection="best",
)

# Learn equations
model = PySRRegressor(
    niterations=1000,
    binary_operators=["plus", "mult"],
    unary_operators=["cos", "sin",],
    nested_constraints={"sin": {"cos": 0}, "sin": {"sin": 0}, "cos": {"sin": 0}, "cos": {"cos": 0}},
    **default_pysr_params
)

model.fit(X, y)
```

**Reply by MilesCranmer** [reply]
Increase the maxsize so it can find it. Also check out the tuning page: https://astroautomata.com/PySR/tuning/.

By the way, in your code, you need this change:

```diff
- nested_constraints={"sin": {"cos": 0}, "sin": {"sin": 0}, "cos": {"sin": 0}, "cos": {"cos": 0}},
+ nested_constraints={"sin": {"cos": 0, "sin": 0}, "cos": {"sin": 0, "cos": 0}},
```

**Reply by MilesCranmer** [reply, marked as answer]
It found your equation for me with these settings:

```py
model = PySRRegressor(
    niterations=1000,
    maxsize=40,
    binary_operators=["+", "-", "*"],
    unary_operators=["sin", "square"],
    nested_constraints={"sin": {"sin": 0}, "square": {"sin": 0}},
    constraints={"sin": 8, "square": 2},
    parsimony=0.01,
    **default_pysr_params
)
model.fit(X, y)
```
it found it maybe 20% of the way through.
